In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('dataset/Feature_vectors_by_customerID_and_month.tsv', sep='\t')
df.head()

CustomerID  Month  No_planned_visits  No_executed_visits  \
0   100000136      5                  7                   7   
1   100000136      6                  8                   8   
2   100000136      7                  4                   4   
3   100000136      8                  8                   8   
4   100000136      9                  9                   9   

   No_not_executed_visits  No_adhoc_visits  Day_maximum_difference  \
0                       0                0                     8.0   
1                       0                0                    15.0   
2                       0                0                    24.0   
3                       0                0                     3.0   
4                       0                0                     4.0   

   Day_maximum_sape  Day_minimum_sape  Day_validation       ...        \
0               4.0               4.0            12.0       ...         
1              15.0              29.0             NaN       ...         
2              24.0               3.0             NaN       ...         
3              31.0              31.0             NaN       ...         
4              31.0              31.0             NaN       ...         

   Mean_SAPE  Std_dev_SAPE  Mean_APE  Squared_error_SAPE_APE  Mean_FPE  \
0    90.9870  1.421085e-14    79.498              131.997121    90.987   
1    92.6325  1.744089e+00       NaN                     NaN    92.633   
2    95.3750  6.250000e-01       NaN                     NaN    95.375   
3    96.0000  0.000000e+00       NaN                     NaN    96.000   
4    96.0000  0.000000e+00       NaN                     NaN    96.000   

    Std_dev_FPE  No_ok_validations  No_not_ok_validations  Percent_ok  \
0  1.421085e-14                  1                      0         1.0   
1  1.421085e-14                  0                      0         0.0   
2  0.000000e+00                  0                      0         0.0   
3  0.000000e+00                  0                      0         0.0   
4  0.000000e+00                  0                      0         0.0   

   Percent_not_ok  
0             0.0  
1             0.0  
2             0.0  
3             0.0  
4             0.0  

[5 rows x 21 columns]

In [34]:
no_records_wno_validation = len(df[df.No_validations == 0])
print(no_records_wno_validation)

30254


За 30254 записи не постои валидација, односно атрибутот No_validations има вредност 0. Бидејќи не постои валидација, ова повлекува дека и атрибутите Mean_APE, Squared_error_SAPE_APE и Day_validation ќе имаат np.nan вредности.
Nan вредностите на овие атрибути ќе бидат заменети со -1. 

In [31]:
mean_ape_no_nan = np.zeros(len(df))
sqrd_err_no_nan = np.zeros(len(df))
day_of_validation_no_nan = np.zeros(len(df))
    
for i in range(len(df)):
    mean_ape = df['Mean_APE'][i]
    sqrd_err = df['Squared_error_SAPE_APE'][i]
    day_of_validation = df['Day_validation'][i]
    
    mean_ape_no_nan[i] = -1 if np.isnan(mean_ape) else mean_ape
    sqrd_err_no_nan[i] = -1 if np.isnan(sqrd_err) else sqrd_err
    day_of_validation_no_nan[i] = -1 if np.isnan(day_of_validation) else day_of_validation

df.Mean_APE = pd.Series(mean_ape_no_nan, dtype=float)
df.Squared_error_SAPE_APE = pd.Series(sqrd_err_no_nan, dtype=float)
df.Day_validation = pd.Series(day_of_validation_no_nan, dtype=int)

In [48]:
df[(df.No_executed_visits == 0) & (df.No_adhoc_visits == 0) & (~np.isnan(df.Mean_SAPE))]

CustomerID  Month  No_planned_visits  No_executed_visits  \
18007   100007200      7                  0                   0   
32473   100023111      7                  1                   0   

       No_not_executed_visits  No_adhoc_visits  Day_maximum_difference  \
18007                       0                0                     2.0   
32473                       1                0                     2.0   

       Day_maximum_sape  Day_minimum_sape  Day_validation       ...        \
18007               2.0               2.0              -1       ...         
32473               2.0               2.0              -1       ...         

       Mean_SAPE  Std_dev_SAPE  Mean_APE  Squared_error_SAPE_APE  Mean_FPE  \
18007     33.903           0.0      -1.0                    -1.0    33.903   
32473     71.181           0.0      -1.0                    -1.0    71.181   

       Std_dev_FPE  No_ok_validations  No_not_ok_validations  Percent_ok  \
18007          0.0                  0                      0         0.0   
32473          0.0                  0                      0         0.0   

       Percent_not_ok  
18007             0.0  
32473             0.0  

[2 rows x 21 columns]

In [49]:
df[(df.No_executed_visits == 0) & (df.No_adhoc_visits == 0) & (~np.isnan(df.Mean_FPE))]

CustomerID  Month  No_planned_visits  No_executed_visits  \
18007   100007200      7                  0                   0   
32473   100023111      7                  1                   0   

       No_not_executed_visits  No_adhoc_visits  Day_maximum_difference  \
18007                       0                0                     2.0   
32473                       1                0                     2.0   

       Day_maximum_sape  Day_minimum_sape  Day_validation       ...        \
18007               2.0               2.0              -1       ...         
32473               2.0               2.0              -1       ...         

       Mean_SAPE  Std_dev_SAPE  Mean_APE  Squared_error_SAPE_APE  Mean_FPE  \
18007     33.903           0.0      -1.0                    -1.0    33.903   
32473     71.181           0.0      -1.0                    -1.0    71.181   

       Std_dev_FPE  No_ok_validations  No_not_ok_validations  Percent_ok  \
18007          0.0                  0                      0         0.0   
32473          0.0                  0                      0         0.0   

       Percent_not_ok  
18007             0.0  
32473             0.0  

[2 rows x 21 columns]

Отстранување на редиците за кои што атрибутите No_executed_visits и No_adhoc_visits имаат вредност 0. Ова значи дека во даден објект (CustomerID), за даден месец (Month) не е извршена ниту планирана, ниту адхок посета. 

Постојат само 2 записи за кои што овие два атрибути имаат вредност 0, а вредноста на Mean_SAPE или Mean_FPE е различна од np.nan.
Овие записи се најверојатно грешка, бидејќи во тој месец во објектот не е направена никаква посета од страна на комерцијалист, а постојат коефициенти за евалуација на анкета.

По ова чистење, може да се отстрани и атрибутот No_validations (вредности 0 и 1), бидејќи секогаш кога овој атрибут има вредност 0, Day_validation има вредност -1, а секогаш кога овој атрибут има вредност 1, Day_validation е позитивен цел број. 

In [59]:
df = df.drop(df.index[(df.No_executed_visits == 0) & (df.No_adhoc_visits == 0)], axis=0)
df = df.drop(['No_validations'], axis=1)

In [106]:
strange_customers = df[(df.No_executed_visits > 0) & (np.isnan(df.Mean_SAPE))]
strange_customers.to_csv('subsets/Customers_w_executed_visits_and_no_sape.tsv', sep='\t', index=None)
len(strange_customers)

65

Претходните записи се неконзистентни. За даден објект и месец, забележано е дека посета (или повеќе посети) од страна на комерцијалист е извршена, но не е забележан никаков SAPE коефициент.

Претходните 65 записи ќе бидат отстранети бидејќи немаат никакво значење.

In [108]:
df = df.drop(df.index[(df.No_executed_visits > 0) & (np.isnan(df.Mean_SAPE))], axis=0)
len(df)

32197

In [110]:
not all(map(np.isnan, df.Mean_FPE))  # нема ниту еден запис со MEAN_FPE = np.nan

True

In [111]:
not all(map(np.isnan, df.Mean_SAPE)) # нема ниту еден запис со MEAN_SAPE = np.nan

True

In [109]:
df.to_csv('dataset/Feature_vectors_customer_month_cleaned.tsv', sep='\t', index=None)